cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -198,17 +198,22 @@ static util::optional<compute::Expression> ColumnChunkStatisticsAsExpression(
   auto maybe_min = min->CastTo(field->type());
   auto maybe_max = max->CastTo(field->type());
   if (maybe_min.ok() && maybe_max.ok()) {
-    auto col_min = maybe_min.MoveValueUnsafe();
-    auto col_max = maybe_max.MoveValueUnsafe();
-    if (col_min->Equals(col_max)) {
-      return compute::equal(std::move(field_expr), compute::literal(std::move(col_min)));
+    min = maybe_min.MoveValueUnsafe();
+    max = maybe_max.MoveValueUnsafe();
+    if (min->Equals(max)) {
+      return compute::equal(std::move(field_expr), compute::literal(std::move(min)));
```

Comment:
Do we want to add an `is_valid` here too if applicable?

Reply:
Equality expressions will be picked up by the ExtractKnownFieldValues pass, which assumes the field will `min` and never null (questionably correct)

Longer term: I think we need to be more rigorous with the guarantee contract. Specifically: any guarantee should be usable as a filter, and in that case will evaluate to an array containing only `true` (and no nulls).

| Data, Guarantee | Filter | Evaluation of guarantee on Data | |
|---|---|---|---|
| [2, 2], i32 == 2 | i32 == 2 | [true, true] | can simplify to `true` |
| [2, null], i32 == 2 or is_null(i32) | i32 == 2 | [true, null] | the second row should be excluded from the filter |

Reply:
Sounds good to me. I think a row group with `[2, null]` right now might actually get filtered out accidentally if your filter is `i32 != 2`, since the guarantee will simply be `i32 == 2`…